include/ensmallen_bits/callbacks/report.hpp

```diff
@@ -0,0 +1,558 @@
+   *     MaxIterations().
+   * @param stream The output stream.
+   */
+  template<typename OptimizerType>
+  typename std::enable_if<
+      traits::HasMaxIterationsSignature<OptimizerType>::value, void>::type
+  PrintMaxIterations(const OptimizerType& optimizer, std::stringstream& stream)
+  {
+    PrettyPrintElement(stream, "Max iterations:", 30);
+    stream << optimizer.MaxIterations() << std::endl;
+
+    PrettyPrintElement(stream, "Reached maximum iterations:", 30);
+    stream << std::string(optimizer.MaxIterations() == objectives.size() ?
+        "true" : "false") << std::endl;
+  }
+
+  template<typename OptimizerType>
+  typename std::enable_if<
+      !traits::HasMaxIterationsSignature<OptimizerType>::value, void>::type
+  PrintMaxIterations(const OptimizerType& /* optimizer */,
+                     std::stringstream& /* stream */) { }
+
+  /**
+   * Helper function to output the batch-size to the specified output stream.
+   *
+   * @param optimizer The instantiated optimizer that implements BatchSize().
+   * @param stream The output stream.
+   */
+  template<typename OptimizerType>
+  typename std::enable_if<traits::HasBatchSizeSignature<OptimizerType>::value,
+      void>::type
+  PrintBatchSize(const OptimizerType& optimizer, std::stringstream& stream)
+  {
+    PrettyPrintElement(stream, "Batchsize:", 30);
+    stream << optimizer.BatchSize() << std::endl;
+  }
+
+  template<typename OptimizerType>
+  typename std::enable_if<!traits::HasBatchSizeSignature<OptimizerType>::value,
+      void>::type
+  PrintBatchSize(const OptimizerType& /* optimizer */,
+                 std::stringstream& /* stream */) { }
+
+  /**
+   * Output formated data.
+   *
+   * @param out Output stream.
+   * @param data The data to print on the given stream.
+   * @param width The width of the the formated output data.
+   */
+  template<typename T>
+  void PrettyPrintElement(std::ostream& out,
+                          const T& data,
+                          const size_t width = 14)
+  {
+    out << std::left << std::setw(width) << std::setfill(' ')
+        << std::setprecision(3) << data;
+  }
+
+  /**
+   * Output formated data.
+   *
+   * @param data The data to print on the given stream.
+   * @param width The width of the the formated output data.
+   */
+  template<typename T>
+  void PrettyPrintElement(const T& data, const size_t width = 14)
+  {
+    PrettyPrintElement(output, data, width);
+  }
+
+  /**
+   * Outputs the given matrix in a truncated format. For example, the matrix:
+   *
+   * 1 2 3 4 5
+   * 6 7 8 9 10
+   * 11 12 13 14
+   * 15 16 17 18
+   *
+   * will be truncated to:
+   *
+   * 1 2 ... 5
+   * 6 7 ... 10
+   * ...
+   * 15 16 ... 18
+   *
+   * @param data The data to print on the given stream in a truncated format.
+   * @param size The number of elements per column/row.
+   */
+  template<typename T>
+  void TruncatePrint(const T& data, const size_t size)
```

Comment:
This is a nice function!

@conradnicta do you think that you might want to merge this back upstream into Armadillo?  I'm not sure how exactly it would fit, but it might be useful in addition to `.print()` (or maybe as an extra option to `.print()` or something).  Anyway, no opinion from my side, I just thought it might be useful to bring it to your attention. :)

Reply:
@rcurtin Sorry, only saw this now.  The @ was missing an 's'.
A version of .print() which prints an abridged matrix would indeed be useful -- similar to Python.  I'll add that on the TODO list for Armadillo 10.2

Reply:
Oh, you're right, I missed a key.  Sorry about that!

Reply:
@rcurtin @zoq I've added `.brief_print()` to Armadillo 10.2: http://arma.sourceforge.net/docs.html#brief_print

Reply:
This is nice.